core/solver/cb_gmres.cpp

```diff
@@ -188,320 +189,343 @@ struct helper<std::complex<T>> {
-        }
-        // Start only after this value with performing forced iterations after
-        // convergence detection
-        constexpr decltype(total_iter) start_force_reset{10};
-        bool perform_reset{false};
-        // Fraction of the krylov_dim_ (or total_iter if it is lower),
-        // determining the number of forced iteration to perform
-        constexpr decltype(krylov_dim_) forced_iteration_fraction{10};
-        const decltype(krylov_dim_) forced_limit{krylov_dim_ /
-                                                 forced_iteration_fraction};
-        // Counter for the forced iterations. Start at max in order to properly
-        // test convergence at the beginning
-        decltype(krylov_dim_) forced_iterations{forced_limit};
-
-        while (true) {
-            ++total_iter;
-            this->template log<log::Logger::iteration_complete>(
-                this, total_iter, residual.get(), dense_x, residual_norm.get());
-            // In the beginning, only force a fraction of the total iterations
-            if (forced_iterations < forced_limit &&
-                forced_iterations < total_iter / forced_iteration_fraction) {
-                ++forced_iterations;
-            } else {
-                bool all_changed = stop_criterion->update()
-                                       .num_iterations(total_iter)
-                                       .residual(residual.get())
-                                       .residual_norm(residual_norm.get())
-                                       .solution(dense_x)
-                                       .check(RelativeStoppingId, true,
-                                              &stop_status, &one_changed);
-                if (one_changed || all_changed) {
-                    host_stop_status = stop_status;
-                    bool host_array_changed{false};
-                    for (size_type i = 0; i < host_stop_status.get_num_elems();
-                         ++i) {
-                        auto local_status = host_stop_status.get_data() + i;
-                        // Ignore all actually converged ones!
-                        if (fully_converged_rhs.get_data()[i]) {
-                            continue;
-                        }
-                        if (local_status->has_converged()) {
-                            // If convergence was detected earlier, or
-                            // at the very beginning:
-                            if (stop_encountered_rhs.get_data()[i] ||
-                                total_iter < start_force_reset) {
-                                fully_converged_rhs.get_data()[i] = true;
+    precision_dispatch_real_complex<ValueType>(
+        [&](auto dense_b, auto dense_x) {
+            // Current workaround to get a lambda with a template argument (only
+            // the type of `value` matters, the content does not)
+            auto apply_templated = [&](auto value) {
```

Comment:
I should look into an alternative, we have quite a lot of blocks opened here, which makes the readability quite hard (especially following the indentations). But for this PR, it is totally fine.

Reply:
I had some similar thoughts, the only reason I need to use the Lambda here is that I wanted to support real-to-complex apply, which involves a temporary object I need to copy back to, and convert_back_deleter only stores a pointer. Maybe I should add a copy_back_temporary_deleter?

Reply:
I meant the surrounding Lambda, your code is fine :) copy_back_temporary_deleter would store a unique_ptr to the temporary instead of a raw pointer to the original object.

Reply:
I don't see how this would impact this code, you would still need the lambda function in order to have the correct type unless you plug it directly into `linop.hpp`.

Reply:
you can see make_temporary_conversion like a more complex as<Dense<...>> call, the static type of dense_x and dense_b is always Dense<ValueType>

Reply:
Ah, you mean instead of having the surrounding lambda, you would change the `as<Dense<ValueType>>` calls with a `make_temporary_conversion<Dense<ValueType>>` call?

Reply:
Yes, that's what I had in place previously, but that didn't support real-to-complex apply yet, but hopefully will soon.

Reply:
I think the current solution is fine, and since time is slightly pressing, I would prefer to merge the current (working) solution and update it if we feel the need / have a working, better alternative.